server/app/controllers/applicant/ApplicantProgramReviewController.java

```diff
@@ -169,7 +170,12 @@ private CompletionStage<Result> submitInternal(
             programId,
             submittingProfile,
             featureFlags.isProgramEligibilityConditionsEnabled(request));
-    return submitApp
+    CompletableFuture<ReadOnlyApplicantProgramService> readOnlyApplicantProgramServiceFuture =
+        applicantService
+            .getReadOnlyApplicantProgramService(applicantId, programId)
+            .toCompletableFuture();
+    return readOnlyApplicantProgramServiceFuture
+        .thenComposeAsync(v -> submitApp.toCompletableFuture())
         .thenApplyAsync(
```

Comment:
**thought (non-blocking):** this does the right thing I think. AFAICT the tricky thing here is having multiple futures we want to resolve before continuing, another way to do that could be to use `CompleteableFuture.allOf`, which may make that intention a bit more clear (I always forget what `thenCompose` does other than it's usually not what I expect or want).

I don't know if using `allOf` is actually an improvement for readability though so up to you.

Reply:
Good call - its somewhat similar for readability I think. I used allOf instead of what I had before, since these don't depend on one another - let me know if you have a preference for the other method though.

Reply:
LGTM!